File: solutions/Virtual-Assistant/src/csharp/assistant/Dialogs/Escalate/EscalateDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using VirtualAssistant.Dialogs.Shared;

namespace VirtualAssistant.Dialogs.Escalate
{
    public class EscalateDialog : EnterpriseDialog
    {
        private EscalateResponses _responder = new EscalateResponses();

        public EscalateDialog(BotServices botServices, IBotTelemetryClient telemetryClient)
            : base(botServices, nameof(EscalateDialog), telemetryClient)
        {
            InitialDialogId = nameof(EscalateDialog);
            TelemetryClient = telemetryClient;

            var escalate = new WaterfallStep[]
            {
                SendEscalationMessage,
            };

            AddDialog(new WaterfallDialog(InitialDialogId, escalate) { TelemetryClient = telemetryClient });
        }

        private async Task<DialogTurnResult> SendEscalationMessage(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            await _responder.ReplyWith(sc.Context, EscalateResponses.ResponseIds.SendEscalationMessage);
            return await sc.EndDialogAsync();
        }
    }
}

File: solutions/Virtual-Assistant/src/typescript/assistant/src/dialogs/escalate/escalateDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { BotTelemetryClient } from 'botbuilder';
import {
        DialogTurnResult,
        WaterfallDialog,
        WaterfallStepContext } from 'botbuilder-dialogs';
import { BotServices } from '../../botServices';
import { EnterpriseDialog } from '../shared/enterpriseDialog';
import { EscalateResponses } from './escalateResponses';

export class EscalateDialog extends EnterpriseDialog {
    // Fields
    private static readonly responder: EscalateResponses = new EscalateResponses();

    // Constructor
    constructor(botServices: BotServices, telemetryClient: BotTelemetryClient) {
        super(botServices, EscalateDialog.name, telemetryClient);
        this.initialDialogId = EscalateDialog.name;
        const escalate: ((sc: WaterfallStepContext<{}>) => Promise<DialogTurnResult>)[] = [
            EscalateDialog.sendEscalationMessage.bind(this)
        ];
        this.addDialog(new WaterfallDialog(this.initialDialogId, escalate));
    }

    private static async sendEscalationMessage(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        await EscalateDialog.responder.replyWith(sc.context, EscalateResponses.responseIds.sendEscalationMessage);

        return sc.endDialog();
    }
}
